<a href="https://colab.research.google.com/github/Kridr/HW_PB_HSE/blob/main/HW1_PB.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Выполнил: Ляликов Никита, мИИАД22

Установка необходимой библиотеки

In [ ]:
!pip install biopython

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 2.3 MB 4.7 MB/s 


Импортирование библиотеки Bio

In [ ]:
import Bio
from Bio import Entrez
from Bio import SeqIO
print(Bio.__version__)

1.79


Считывание нужной нам бактерии. Поставим alias для бактерии. Эта та же самая бактерия но из другой БД (Оригинальное название находится в RefSeq, а это в GenBank). Если брать оригинальное название, то последовательность пустая (то есть поле seq в объекте SeqRecord имеет значение None)

In [ ]:
handle = Entrez.efetch(db="nucleotide", rettype="gb", retmode="text",
                       id="CP063829.1")
for seq_record in SeqIO.parse(handle, "gb"):
    print (seq_record.id, seq_record.description[:50] + "...")
    print ("Sequence length %i," % len(seq_record))
    print ("%i features," % len(seq_record.features))
    print ("from: %s" % seq_record.annotations["source"])

/usr/local/lib/python3.7/dist-packages/Bio/Entrez/__init__.py:670: UserWarning: 
Email address is not specified.

To make use of NCBI's E-utilities, NCBI requires you to specify your
email address with each request.  As an example, if your email address
is A.N.Other@example.com, you can specify it as follows:
   from Bio import Entrez
   Entrez.email = 'A.N.Other@example.com'
In case of excessive usage of the E-utilities, NCBI will attempt to contact
a user at the email address provided before blocking access to the
E-utilities.
  UserWarning,


CP063829.1 Streptococcus pneumoniae strain 19A-ST320_99-176 c...
Sequence length 2089994,
4288 features,
from: Streptococcus pneumoniae


In [ ]:
seq_record

SeqRecord(seq=Seq('AAAAACCATTATACAAAATGGAATTTCGTTTTAGATAAAATTCTCTCAACTGTC...TAC'), id='CP063829.1', name='CP063829', description='Streptococcus pneumoniae strain 19A-ST320_99-176 chromosome, complete genome', dbxrefs=['BioProject:PRJNA672899', 'BioSample:SAMN16580452'])

Запишем считанную бактерию в файл fasta

In [ ]:
with open("strept_pneum.fasta", "w") as output_handle:
    SeqIO.write(seq_record, output_handle, "fasta")

Скачаем, распакауем и установим EMBOSS

In [ ]:
!wget ftp://emboss.open-bio.org/pub/EMBOSS/emboss-latest.tar.gz

--2022-09-25 20:15:46--  ftp://emboss.open-bio.org/pub/EMBOSS/emboss-latest.tar.gz
           => ‘emboss-latest.tar.gz’
Resolving emboss.open-bio.org (emboss.open-bio.org)... 52.45.115.20
Connecting to emboss.open-bio.org (emboss.open-bio.org)|52.45.115.20|:21... connected.
Logging in as anonymous ... Logged in!
==> SYST ... done.    ==> PWD ... done.
==> TYPE I ... done.  ==> CWD (1) /pub/EMBOSS ... done.
==> SIZE emboss-latest.tar.gz ... 117962028
==> PASV ... done.    ==> RETR emboss-latest.tar.gz ... done.
Length: 117962028 (112M) (unauthoritative)

emboss-latest.tar.g 100%[===================>] 112.50M  75.4MB/s    in 1.5s    

2022-09-25 20:15:48 (75.4 MB/s) - ‘emboss-latest.tar.gz’ saved [117962028]



In [ ]:
!gunzip emboss-latest.tar.gz

In [ ]:
%%capture
!tar xvf emboss-latest.tar

In [ ]:
%%bash
cd ./EMBOSS-6.6.0/
./configure

checking for gawk... no
checking for mawk... mawk
checking for icc... no
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for icpc... no
checking for g++... g++
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking how to run the C preprocessor... gcc -E
checking for a BSD-compatible install... /usr/bin/install -c
checking whether ln -s works... yes
checking whether make sets $(MAKE)... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking whether build environment is sane... yes
checking for style of include used by make... GNU
checking dependency style of gcc... 

./configure: line 6806: /usr/bin/file: No such file or directory


In [ ]:
%%bash
cd ./EMBOSS-6.6.0/
make

Making all in plplot
make[1]: Entering directory '/content/EMBOSS-6.6.0/plplot'
Making all in lib
make[2]: Entering directory '/content/EMBOSS-6.6.0/plplot/lib'
make[2]: Nothing to be done for 'all'.
make[2]: Leaving directory '/content/EMBOSS-6.6.0/plplot/lib'
make[2]: Entering directory '/content/EMBOSS-6.6.0/plplot'
make[2]: Nothing to be done for 'all-am'.
make[2]: Leaving directory '/content/EMBOSS-6.6.0/plplot'
make[1]: Leaving directory '/content/EMBOSS-6.6.0/plplot'
Making all in ajax
make[1]: Entering directory '/content/EMBOSS-6.6.0/ajax'
Making all in pcre
make[2]: Entering directory '/content/EMBOSS-6.6.0/ajax/pcre'
make[2]: Nothing to be done for 'all'.
make[2]: Leaving directory '/content/EMBOSS-6.6.0/ajax/pcre'
Making all in expat
make[2]: Entering directory '/content/EMBOSS-6.6.0/ajax/expat'
make[2]: Nothing to be done for 'all'.
make[2]: Leaving directory '/content/EMBOSS-6.6.0/ajax/expat'
Making all in zlib
make[2]: Entering directory '/content/EMBOSS-6.6.0/ajax/zlib'

ar: `u' modifier ignored since `D' is the default (see `U')
ar: `u' modifier ignored since `D' is the default (see `U')
ar: `u' modifier ignored since `D' is the default (see `U')
ar: `u' modifier ignored since `D' is the default (see `U')


Последовательно выполняем задание

Разделим последовательность на 10 частей, запишем в новый файл

In [ ]:
!./EMBOSS-6.6.0/emboss/splitter -sequence strept_pneum.fasta -size 10 -outseq strept_pheum10.fasta

Split sequence(s) into smaller sequences


Посчитаем частоты всех последовательностей длины 4

In [ ]:
!./EMBOSS-6.6.0/emboss/compseq -sequence strept_pheum10.fasta -word 4 -outfile first.txt

Calculate the composition of unique words in sequences


Выпишем информацию об интересующих последовательностях

In [ ]:
with open ('first.txt', 'r') as first:
    for line in first:
        if line.startswith('AAAA') or line.startswith('ATAT') or line.startswith('ATTA') or line.startswith('AATT') or line.startswith('# Word'):
            print(line)

# Word	Obs Count	Obs Frequency	Exp Frequency	Obs/Exp Frequency

AAAA	19979		0.0136562	0.0039062	3.4959955

AATT	10879		0.0074361	0.0039062	1.9036456

ATAT	7913		0.0054088	0.0039062	1.3846445

ATTA	6983		0.0047731	0.0039062	1.2219098



Перемешаем последовательности

In [ ]:
!./EMBOSS-6.6.0/emboss/shuffleseq -sequence strept_pheum10.fasta -outseq strept_pheum10_shuf.fasta

Shuffle a set of sequences maintaining composition


Опять посчитаем и выведем интресующие данные

In [ ]:
!./EMBOSS-6.6.0/emboss/compseq -sequence strept_pheum10_shuf.fasta -word 4 -outfile second.txt

Calculate the composition of unique words in sequences


In [ ]:
with open ('second.txt', 'r') as second:
    for line in second:
        if line.startswith('AAAA') or line.startswith('ATAT') or line.startswith('ATTA') or line.startswith('AATT') or line.startswith('# Word'):
            print(line)

# Word	Obs Count	Obs Frequency	Exp Frequency	Obs/Exp Frequency

AAAA	15579		0.0106487	0.0039062	2.7260681

AATT	10210		0.0069788	0.0039062	1.7865816

ATAT	10270		0.0070198	0.0039062	1.7970806

ATTA	10307		0.0070451	0.0039062	1.8035550



По итогу получается следующее:

До перемешивания:

Word	Obs Count	Obs Frequency	Exp Frequency	Obs/Exp Frequency

AAAA	19979		0.0136562	0.0039062	3.4959955

AATT	10879		0.0074361	0.0039062	1.9036456

ATAT	7913		0.0054088	0.0039062	1.3846445

ATTA	6983		0.0047731	0.0039062	1.2219098

---
После перемешивания

Word	Obs Count	Obs Frequency	Exp Frequency	Obs/Exp Frequency

AAAA	15653		0.0106993	0.0039062	2.7390169

AATT	10261		0.0070137	0.0039062	1.7955058

ATAT	10182		0.0069597	0.0039062	1.7816821

ATTA	10119		0.0069166	0.0039062	1.7706581


Как видно, для всех слов в обоих наблюдаемая частота ниже, чем ожидаемая. Это можно увидеть по отношению в последнем столбце

При этом, для слов AAAA и AATT отношение в последнем столбце выше для первого случая (до перемешивания), а для слов ATATи ATTA - для второго случая (после перемешивания)

Я не уверен, что проделал все правильно, поэтому хочу проделать все в Python вручную для проверки

Разделим последовательности на 10 частей

In [ ]:
str_seq = str(seq_record.seq)

parts = 10
chunks, chunk_size = len(str_seq), len(str_seq) // parts
str_seq_list = [str_seq[i : i+chunk_size] for i in range(0, chunks, chunk_size)]

Посчитаем частоты как для оригинальных последовательностей, так и для перемешанных

In [ ]:
from collections import defaultdict
import random

counter = defaultdict(int)
counter_shuf = defaultdict(int)

all_pos_count = sum([len(str_seq_list[i]) - 3 for i in range(parts)])

for i in range(parts):
    for word in ['AAAA', 'AATT', 'ATAT', 'ATTA']:
        counter[word] += str_seq_list[i].count(word)

        l = list(str_seq_list[i])
        random.shuffle(l)
        str_seq_list_i_shuffled = ''.join(l)
        counter_shuf[word] += str_seq_list_i_shuffled.count(word)

freq = {(k, v / all_pos_count) for k, v in counter.items()}
freq_shuf = {(k, v / all_pos_count) for k, v in counter_shuf.items()}

In [ ]:
counter

defaultdict(int, {'AAAA': 18963, 'AATT': 15594, 'ATAT': 10865, 'ATTA': 9709})

In [ ]:
counter_shuf

defaultdict(int, {'AAAA': 11871, 'AATT': 16932, 'ATAT': 15639, 'ATTA': 16628})

Совпадения по количеству нет, поэтому продолжать не вижу смысла. Видимо в реальности это должно считаться как-то по-другому